quill-jdbc-monix/src/main/scala/io/getquill/context/jdbc/AsyncJdbcContext.scala

```diff
@@ -0,0 +1,201 @@
+      }
+
+  def transaction[A](f: Task[A]): Task[A] = {
+    val dbEffects = for {
+      result <- currentConnection.get match {
+        case Some(_) => f // Already in a transaction
+        case None =>
+          Task {
+            val connection = dataSource.getConnection
+            val wasAutoCommit = connection.getAutoCommit
+            connection.setAutoCommit(false)
+            (connection, wasAutoCommit)
+          }.bracket {
+            case (connection, _) =>
+              Task(connection).flatMap { connection =>
+                currentConnection.update(Some(connection))
+                f.onCancelRaiseError(new IllegalStateException()) // TODO Change this to another error
+                  .doOnFinish {
+                  case Some(error) =>
+                    connection.rollback()
+                    Task.raiseError(error)
+                  case None =>
+                    Task(connection.commit())
+                }
+              }
+          } {
+            case (connection, wasAutoCommit) =>
+              Task {
+                currentConnection.update(None)
+                connection.setAutoCommit(wasAutoCommit)
+                connection.close()
+              }
+          }
+      }
+    } yield result
+
+    dbEffects
+      .executeOn(transactionScheduler)
+      .executeWithOptions(_.enableLocalContextPropagation)
+      .asyncBoundary
+  }
+
+  override def probe(statement: String): Try[_] = Try {
+    withConnection(conn => Task(conn.createStatement.execute(statement))).runSyncUnsafe()
+  }
+
+  class ResultSetIterator[T](rs:ResultSet, extractor: Extractor[T]) extends BufferedIterator[T] {
+
+    private[this] var state = 0 // 0: no data, 1: cached, 2: finished
+    private[this] var cached: T = null.asInstanceOf[T]
+
+    protected[this] final def finished(): T = {
+      state = 2
+      null.asInstanceOf[T]
+    }
+
+    /** Return a new value or call finished() */
+    protected def fetchNext(): T =
+      if (rs.next()) extractor(rs)
+      else finished()
+
+    def head: T = {
+      prefetchIfNeeded()
+      if(state == 1) cached
+      else throw new NoSuchElementException("head on empty iterator")
+    }
+
+    private def prefetchIfNeeded():Unit = {
+      if(state == 0) {
+        cached = fetchNext()
+        if(state == 0) state = 1
+      }
+    }
+
+    def hasNext: Boolean = {
+      prefetchIfNeeded()
+      state == 1
+    }
+
+    def next(): T = {
+      prefetchIfNeeded()
+      if(state == 1) {
+        state = 0
+        cached
+      } else throw new NoSuchElementException("next on empty iterator");
+    }
+  }
+
+  def streamQuery[T](sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): Observable[T] = {
+    val obs =
+      withConnectionObservable { conn =>
+        Observable.eval {
+          val (params, ps) = prepare(conn.prepareStatement(sql))
```

Comment:
This needs to set some additional properties for the Postgres driver not to actually materialize the whole ResultSet in buffers. See 
https://jdbc.postgresql.org/documentation/head/query.html#query-with-curso. Most notably, auto commit must be set to false, fetchSize may need to be tuned, and may also require setting result set parameters on the statement before execution. I remember having issues running out of memory using slick until I figured out those details.

Reply:
Hi @pettyjamesm. Thanks for the note! I'll be sure to make these adjustments for postgres. Any idea if there's an easier way to test this besides blowing up the heap with a giant table?

Reply:
Come to think of it, maybe `fetchSize` should be part of the `stream` method directly (Slick does this also). It's also interesting to note that other databases e.g. MySQL require similar kinds of parameter tuning (have a look [here](https://stackoverflow.com/questions/24676111/streaming-mysql-resultset-with-fixed-number-of-results-at-a-time)). We probably need a place to put in these streaming-specific vendor-specific settings.

Does this look right?
````scala
protected def withConnectionObservable[T](f: Connection => Observable[T]): Observable[T] =
  for {
    maybeConnection <- Observable { currentConnection.get }
    result <- maybeConnection match {
      case Some(connection) =>
        // Need to store, set and restore the client's autocommit mode since some vendors (e.g. postgres)
        // don't like autocommit=true during streaming sessions. Using brackets to do that
        Observable(NonAutoConnectionWrapper(connection))
          .bracket(wrap => f(wrap.conn))(wrap => Task(wrap.restore())).executeOn(transactionScheduler)
      case None =>
        Observable(NonAutoConnectionWrapper(dataSource.getConnection))
          .bracket(wrap => f(wrap.conn))(wrap => Task(wrap.restoreClose())).executeOn(transactionScheduler)
    }
  } yield result

object NonAutoConnectionWrapper {
  def apply(conn:Connection): NonAutoConnectionWrapper = {
    val wasAutoCommit = conn.getAutoCommit
    conn.setAutoCommit(false)
    NonAutoConnectionWrapper(conn, wasAutoCommit)
  }
}
````

Reply:
I can't think of a way off the top of my head, you might be able to manually verify by stalling the processing of the result set and then querying the DB for active connection states. In postgres I would expect to see the streaming connection as `"idle in transaction"`, not sure what info you can see from MySQL.

Another point worth noting, is that calls to `conn.setAutoCommit(...)` both in the connection acquire and connection close can throw an exception but you still need to call `.close()` so that the connection is returned to the connection pool (whereupon it would likely be detected to be dead and replaced).

Final note: we've gotten a lot of mileage from being able to override `executeQuery`, `executeAction` in our customized `PostgresJdbcContext` to wrap a call to `super.executeXXX` with timing information and recording the query being executed. Obviously in a streamed query that won't be so simple, but you might consider adding the query start and finish calls through an overridable protected method so that it's possible for users interested in similar capabilities.

Reply:
@pettyjamesm Okay, I’ll treat a no-autocommit -> (action) -> restore-prev-autocommit chain as a new bracket inside the connection open/close bracket. 

Also, @oleg-py had the idea of intoducing a client-overrideable ‘delay’ method which wraps things inside a Task for every `execute...` method. The default implementation of this is `Task.eval` but the client can potentially use this with things like `Task.shift` to introduce a async boundary. I’m trying to understand, if there is an async boundary being introduced, shouldn’t it come before the connection is aquired? (I.e. it should wrap `currentConnection.get` as opposed to wrapping tasks in later stages? @cemelo, please feel free to comment :)

Also @pettyjamessm. Having connection tracking/audit capability is an awesome idea, but I think it belongs in a separate issue/PR.